include/mbedtls/eddsa.h

```diff
@@ -0,0 +1,285 @@
+ *                  is invalid.
+ * \return          An \c MBEDTLS_ERR_ECP_XXX or \c MBEDTLS_MPI_XXX
+ *                  error code on failure for any other reason.
+ */
+int mbedtls_eddsa_verify( mbedtls_ecp_group *grp,
+                          const unsigned char *buf, size_t blen,
+                          const mbedtls_ecp_point *Q, const mbedtls_mpi *r,
+                          const mbedtls_mpi *s,
+                          mbedtls_eddsa_id eddsa_id,
+                          const unsigned char *ed_ctx, size_t ed_ctx_len );
+
+/**
+ * \brief           This function computes the EdDSA signature and writes it
+ *                  to a buffer, serialized as defined in <em>RFC-4492:
+ *                  Elliptic Curve Cryptography (ECC) Cipher Suites for
+ *                  Transport Layer Security (TLS)</em>.
+ *
+ * \warning         It is not thread-safe to use the same context in
+ *                  multiple threads.
+ *
+ * \note            If the bitlength of the message hash is larger than the
+ *                  bitlength of the group order, then the hash is truncated as
+ *                  defined in <em>Standards for Efficient Cryptography Group
+ *                  (SECG): SEC1 Elliptic Curve Cryptography</em>, section
+ *                  4.1.3, step 5.
+ *
+ * \see             ecp.h
+ *
+ * \param ctx       The ECP keypair context to use. This must be initialized
+ *                  and have a group and private key bound to it, for example
+ *                  via mbedtls_eddsa_genkey().
+ * \param hash      The message hash to be signed. This must be a readable
+ *                  buffer of length \p blen Bytes.
+ * \param hlen      The length of the hash \p hash in Bytes.
+ * \param sig       The buffer to which to write the signature.
+ *                  A buffer size of #MBEDTLS_EDDSA_MAX_LEN is always safe.
+ * \param sig_size  The size of the \p sig buffer in bytes.
+ * \param slen      The address at which to store the actual length of
+ *                  the signature written. Must not be \c NULL.
+ * \param eddsa_id  The signature operation id that identifies PureEdDSA,
+ *                  EdDSActx or EdDSAph
+ * \param ed_ctx    The context for EdDSActx and EdDSAph operations.
+                    it can be \c NULL if \c MBEDTLS_EDDSA_PURE is used or
+                    if no context is provided.
+ * \param ed_ctx_len The length of the context for EdDSActx and EdDSAph
+ *                  operations. It can be \c 0.
+ * \param f_rng     The RNG function. This must not be \c NULL.
+ * \param p_rng     The RNG context to be passed to \p f_rng. This may be
+ *                  \c NULL if \p f_rng is \c NULL or doesn't use a context.
+ *
+ * \return          \c 0 on success.
+ * \return          An \c MBEDTLS_ERR_ECP_XXX, \c MBEDTLS_ERR_MPI_XXX or
+ *                  \c MBEDTLS_ERR_ASN1_XXX error code on failure.
+ */
+int mbedtls_eddsa_write_signature( mbedtls_ecp_keypair *ctx,
```

Comment:
[An EdDSA signature implementation is unsafe if it allows a different public key to be used](https://github.com/MystenLabs/ed25519-unsafe-libs). To be checked in review:

* Does the implementation of `mbedtls_eddsa_write_signature` validate that the public key part of `ctx` is the correct, canonical value?
* Are there tests validating this?